src/Castle.Facilities.AspNet.WebApi/AspNetWebApiFacility.cs

```diff
@@ -0,0 +1,108 @@
+// You may obtain a copy of the License at
+// 
+//     http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+namespace Castle.Facilities.AspNet.WebApi
+{
+	using System;
+	using System.Web.Http;
+	using System.Web.Http.Dispatcher;
+
+	using Castle.MicroKernel;
+	using Castle.MicroKernel.Facilities;
+
+	public class AspNetWebApiFacility : AbstractFacility, ISupportFacility
+	{
+		private bool isSelfHosted;
+		private bool initialised = false;
+		private HttpConfiguration httpConfiguration = null;
+		private Action<IKernel, Type> beforeControllerResolved;
+		private Action<IKernel, object> afterControllerReleased;
+		private AspNetWebApiControllerActivator controllerActivator;
+		private AspNetWebApiDependencyResolver dependencyResolver;
+
+		public new IKernel Kernel => base.Kernel;
+		public bool AutoCreateLifestyleScopes { get; private set; }
+
+		public AspNetWebApiControllerActivator ControllerActivator => controllerActivator;
+		public AspNetWebApiDependencyResolver DependencyResolver => dependencyResolver;
+
+		protected override void Init()
+		{
+			if (!isSelfHosted)
+			{
+				controllerActivator = new AspNetWebApiControllerActivator(this);
+				(httpConfiguration ?? GlobalConfiguration.Configuration).Services.Replace(typeof(IHttpControllerActivator), controllerActivator);
+			}
+			else
+			{
+				dependencyResolver = new AspNetWebApiDependencyResolver(this);
+				(httpConfiguration ?? GlobalConfiguration.Configuration).DependencyResolver = dependencyResolver;
+			}
+
+			SubscribeBeforeControllerCreatedEvent();
+			SubscribeAfterControllerReleaseEvent();
+
+			initialised = true;
+		}
+
+		public AspNetWebApiFacility BeforeControllerResolved(Action<IKernel, Type> beforeResolvedCallback)
+		{
+			beforeControllerResolved = beforeResolvedCallback;
+			if (initialised) SubscribeBeforeControllerCreatedEvent();
+			return this;
+		}
+
+		public AspNetWebApiFacility AfterControllerReleased(Action<IKernel, object> afterReleasedCallback)
+		{
+			afterControllerReleased = afterReleasedCallback;
+			if (initialised) SubscribeAfterControllerReleaseEvent();
+			return this;
+		}
+
+		public AspNetWebApiFacility UsingConfiguration(HttpConfiguration httpConfiguration)
+		{
+			this.httpConfiguration = httpConfiguration;
+			return this;
+		}
+
+		public AspNetWebApiFacility UsingSelfHosting()
+		{
+			isSelfHosted = true;
+			return this;
+		}
+
+		public AspNetWebApiFacility WithLifestyleScopedPerWebRequest()
+		{
+			AutoCreateLifestyleScopes = true;
+			return this;
+		}
+
+		private void SubscribeBeforeControllerCreatedEvent()
+		{
+			if (beforeControllerResolved != null && controllerActivator != null)
```

Comment:
Same comment as the MVC facility about events, it would be good to do something better because it just feels very messy, the control flow is so hard to follow.

Reply:
This has all been refactored to only call back out to the public API. I should never have used this for scope management and now don't anymore. The callbacks are very useful.